examples/funsAndScopes.browse

```diff
@@ -0,0 +1,205 @@
+inMath ignoreThisString {
+  # In here, we have access to abs
+  abs 4.5
+
+  # Also, while the scope doesn't provide `pow`, we can still use the `pow` rule
+  # here because the rule itself is available in the lexical scope, and we are
+  # in a ruleset that is in the math scope
+  pow 2 8
+}
+
+
+# Known Limitations
+#  1. What about rules that rely on multiple scopes in their ancestory? For
+#     instance, `waitAndVisit`, a rule in the `page` scope that relies on `wait`
+#     (from the page scope) and `visit` (from the browser scope)
+#  2. When a rule accepts more than one RuleSet as an argument and plans to
+#     evaluate them, there are multiple ways the scopes can actually get used.
+#     For example, the `for` rule, which takes 2 rulesets, injects a `test` rule
+#     into the first RuleSet, but doesn't do that into the second RuleSet.
+#     However, both RuleSets need to share their variables so that `i`, defined
+#     by `set i 0` in the iterator RuleSet, is still accessible in the body
+#     RuleSet
+
+# An attempt to create the `waitAndVisit` rule described above and ideas to
+# solve these limitations
+
+# providing multiple scopes that should be available for a rule to work
+rule(scope=[browser,page]) waitAndVisit {
+  # ...
+}
+
+# The definition for the page rule would list multiple runtimes for a RuleSet
+rule(runtimes=[,[browser,page]]) page {
+  bind pattern handler
+  # ...
+  evalRuleSet $handler
+  return nil
+}
+
+# now, if a scopes's rules actually rely on another scope, it can define that
+# using `requred`
+scope advMath {
+  requires math
+  rules {
+    round {
+      bind $x
+      set $a (abs $x)
+      if ($x * 2 >= ($a * 2 + 1)) then {
+        return $a
+      } else {
+        return $a + 1
+      }
+    }
+  }
+}
+
+# any function that actually defines runtimes, can specify multiple 
+
+
+############################################
+############################################
+
+# Explicitly passing the scope
+rule waitAndVisit {
+  bind scope ms url
+  
+  # using is a special rule. While we normally have lexical scoping, `using` is
+  # the escape hatch that lets us explicity live in another scope
+  using $scope {    
+    # inside here, we cannot statically determine what rules and vars are
+    # accessible
+
+    wait $ms
+    visit $url
+  }
+
+  # but this bit can only access what's lexically available
+}
+
+page https://windsor.io {
+  # scope is a std rule that just returns the current scope
+  waitAndVisit ( scope ) 100 https://app.windsor.io
+}
+
+############################################
+############################################
+
+# Defining the specific requirements for a rule. The specific rules will be
+# provided via dependency injection
+
+rule waitAndVisit {
+  # We're defining just the specifc requirements for any scope that wishes to
+  # evaluate this rule. It doesn't matter WHAT scope provides these (whether
+  # it's a page or otherwise), just that it exists
+  expects {
```

Comment:
I like this

Reply:
If we are able to push most of the complexity into defining rules, we can keep most scripts very simple by provided great default functions for everything so that defining a rule isn't needed for basic use